tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/TableauEncoder.scala

```diff
@@ -0,0 +1,424 @@
+package at.forsyte.apalache.tla.pp.temporal
+
+import at.forsyte.apalache.tla.lir._
+
+import javax.inject.Singleton
+import com.typesafe.scalalogging.LazyLogging
+import scalaz.Scalaz.{init => _}
+import at.forsyte.apalache.tla.pp.UniqueNameGenerator
+import at.forsyte.apalache.tla.pp.IrrecoverablePreprocessingError
+import at.forsyte.apalache.tla.pp.temporal.utils.builder
+import at.forsyte.apalache.tla.pp.temporal.ScopedBuilderExtensions._
+import at.forsyte.apalache.tla.pp.temporal.DeclUtils._
+import scala.collection.mutable.HashMap
+import at.forsyte.apalache.tla.lir.oper.TlaTempOper
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.lir.transformations.TransformationTracker
+
+/**
+ * Encodes temporal formulas as invariants.
+ *
+ * @author
+ *   Philip Offtermatt
+ */
+@Singleton
+class TableauEncoder(
+    module: TlaModule,
+    gen: UniqueNameGenerator,
+    loopEnc: LoopEncoder,
+    tracker: TransformationTracker)
+    extends LazyLogging {
+  val levelFinder = new TlaLevelFinder(module)
+  val varNamesToExStrings = new HashMap[String, String]()
```

Comment:
For clarity, I'd prefer `var x = immutable.HashMap` over `val x = mutable.HashMap`.

Reply:
Also, why does this need to be mutable in the first place?